### PR TITLE
[ONNX] Support complex in FX exporter

### PR DESCRIPTION
**Pull request by titaiwangms**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #106140
* __->__ #100554

Previous to the PR, the complex dtype would only fail. This PR keeps torch.fx.Graph with complex dtype, while mapping them to float dtype in torchscript(onnx) graph with real representation.

The change happens in multiple files:

1. `placeholder`: Apply torch.view_as_real() before sending fake tensor to graph building.
2. `call_function`: Fill in TorchScriptTensor dtype and shape with real representation dtype and shape.
3. Registry: Add `is_complex`, and supports complex onnxfunction.
4. Dispatcher: Filter with/out complex onnxfunction before opschema matching, based on the dtype in torch args
5. Test cases: input/output view_as_real for result comparisons.

### CODE REVIEW

File: torch/onnx/_internal/fx/passes/fx_to_onnxscript.py [hunk 1]
```diff
@@ -216,6 +216,17 @@ def _fill_tensor_shape_type(
             onnxscript_value.dtype = torch.int64
         elif isinstance(expected_value, torch.SymFloat):
             onnxscript_value.dtype = torch.float32
+        elif expected_value.dtype in _type_utils.COMPLEX_TO_FLOAT:
+            # Like torch.view_as_real, we flatten complex tensors to real tensors with
+            # additional last dimension of 2
+            onnxscript_value.shape = tuple(
+                [dim if isinstance(dim, int) else None for dim in expected_value.size()]
+                + [2]
+            )
```
Comment (justinchuby) [inline torch/onnx/_internal/fx/passes/fx_to_onnxscript.py:225]:
```suggestion
            onnxscript_value.shape = (
                *[dim if isinstance(dim, int) else None for dim in expected_value.size()],
                2
            )
```

File: torch/onnx/_type_utils.py [hunk 2]
```diff
@@ -359,6 +359,11 @@ def valid_torch_name(torch_name: Union[TorchName, str]) -> bool:
     float: {"tensor(float16)", "tensor(float)", "tensor(double)"},
     bool: {"tensor(int32)", "tensor(int64)", "tensor(bool)"},
 }
+COMPLEX_TO_FLOAT = {
```
Comment (justinchuby) [inline torch/onnx/_type_utils.py:362]: I would add a comment for this mapping

File: torch/onnx/_internal/fx/op_validation.py [hunk 3]
```diff
@@ -194,6 +194,9 @@ def generate_random_tensors(shape: torch.Size, dtype: torch.dtype):
         return torch.where(
             random_numbers > 0.5, torch.tensor(True), torch.tensor(False)
         )
+    if dtype in _type_utils.COMPLEX_TO_FLOAT:
+        # ONNX does not support complex value, but support real representation
+        return torch.randn(shape + (2,), dtype=_type_utils.COMPLEX_TO_FLOAT[dtype])
```
Comment (justinchuby) [inline torch/onnx/_internal/fx/op_validation.py:199]:
```suggestion
        return torch.randn((*shape, 2), dtype=_type_utils.COMPLEX_TO_FLOAT[dtype])
```

**Review by justinchuby — DISMISSED** [review]
Awesome!

File: torch/onnx/_type_utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (justinchuby) [reply torch/onnx/_type_utils.py:362]: Good call

[the remark this answers is not in the document]
Comment (titaiwangms) [reply torch/onnx/_type_utils.py]: https://github.com/pytorch/pytorch/pull/103942

**Review by thiagocrepaldi — CHANGES_REQUESTED** [review]

**Review by thiagocrepaldi — DISMISSED** [review]
lgtm. the remaining comments are not blocker, just suggestions

File: test/onnx/test_fx_op_consistency.py [hunk 4]
```diff
@@ -136,11 +137,18 @@
         # "nn.functional.scaled_dot_product_attention"  non-deterministic
         "scatter_add",
         "scatter_reduce",
+        "stft",
```
Comment (BowenBao) [inline test/onnx/test_fx_op_consistency.py:140]: Should `stft` be put in `COMPLEX_TESTED_OPS`?

Comment (titaiwangms) [reply test/onnx/test_fx_op_consistency.py:140]: Sorry, this PR needs registry updated. I will turn it to draft for now.

File: test/onnx/test_fx_op_consistency.py [hunk 5]
```diff
@@ -618,6 +630,51 @@ def _should_skip_xfail_test_sample(
     return None, None
 
 
+def _run_test_output_match(
+    test_suite: onnx_test_common._TestONNXRuntime,
+    device: str,
+    dtype: torch.dtype,
+    op: opinfo_core.OpInfo,
+):
+    # device is provided by instantiate_device_type_tests, but we only want to run in cpu.
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply test/onnx/test_fx_op_consistency.py:617]: Right, I don't think we need to limit ourselves to cpus here, unlike in torchlib. However, I think it's also ok to assume transitivity in `torch gpu` == `torch cpu` == `onnx cpu`? Unless we want to test ORT cuda here.

File: test/onnx/test_fx_op_consistency.py [hunk 6]
```diff
@@ -595,6 +608,51 @@ def _should_skip_xfail_test_sample(
...
+                model.eval()
+
+                if dtype == torch.float32:
+                    # Relax atol and rtol for float32 based on empirical results
+                    rtol = 1e-5
+                    atol = 2e-5
+                else:
+                    rtol = None
+                    atol = None
+                # Run the test
```
Comment (justinchuby) [inline test/onnx/test_fx_op_consistency.py:650]:
```suggestion
```

Use the tolerance field in OpInfo instead?

File: test/onnx/test_fx_op_consistency.py [hunk 7]
```diff
@@ -627,46 +685,15 @@ class TestOnnxModelOutputConsistency(onnx_test_common._TestONNXRuntime):
...
-                        model, inputs, rtol=rtol, atol=atol
-                    )
+        _run_test_output_match(self, device, dtype, op)
+
+    @common_device_type.ops(
+        [op for op in OPS_DB if op.name in COMPLEX_TESTED_OPS],
+        allowed_dtypes=onnx_test_common.COMPLEX_TYPES,
+    )
+    def test_output_match_complex(self, device: str, dtype: torch.dtype, op):
+        """Test the ONNX exporter."""
```
Comment (justinchuby) [inline test/onnx/test_fx_op_consistency.py:695]: I would expand this docstring a little.

File: torch/onnx/_internal/fx/registration.py [hunk 8]
```diff
@@ -26,12 +26,14 @@ class SymbolicFunction:
     op_full_name: The qualified name of the function. In the form of '<namespace>::<op_name>.<overload>'.
     onnx_function: The symbolic function from torchlib.
     is_custom: Whether the function is a custom function.
+    is_complex: Whether the function is a complex function.
```
Comment (justinchuby) [inline torch/onnx/_internal/fx/registration.py:30]:
```suggestion
```

nit

File: torch/onnx/_internal/fx/op_validation.py [hunk 9]
```diff
@@ -238,6 +238,11 @@ def generate_random_tensors(shape: torch.Size, dtype: torch.dtype):
         return torch.where(
             random_numbers > 0.5, torch.tensor(True), torch.tensor(False)
         )
+    if fx_type_utils.is_torch_complex_dtype(dtype):
+        # ONNX does not support complex value, but support real representation
```
Comment (justinchuby) [inline torch/onnx/_internal/fx/op_validation.py:242]:
```suggestion
        # ONNX does not support complex values but supports their real representations
```

nit: grammar for readability

File: torch/onnx/_internal/fx/registration.py [hunk 10]
```diff
@@ -112,6 +125,7 @@ def register_custom_op(
             op_name: The name of the operator to register.
             overload: The overload of the operator to register. If it's default overload,
                 leave it to None.
+            is_complex: Whether the function is a complex function.
```
Comment (justinchuby) [inline torch/onnx/_internal/fx/registration.py:128]:
```suggestion
            is_complex: Whether the function is a function that handles complex inputs.
```

File: torch/onnx/_internal/fx/onnxfunction_dispatcher.py [hunk 11]
```diff
@@ -133,6 +138,64 @@ def dispatch(
             diagnostic_context,
         )
 
+    @_beartype.beartype
+    @diagnostics.diagnose_call(
+        diagnostics.rules.find_operator_overloads_in_onnx_registry,
```
Comment (titaiwangms) [inline torch/onnx/_internal/fx/onnxfunction_dispatcher.py:143]: @BowenBao Does it make sense to have two functions calling the same rule?

File: torch/onnx/_internal/fx/type_utils.py [hunk 12]
```diff
@@ -30,9 +30,11 @@ def dtype(self) -> Optional[torch.dtype]:
         ...
 
 
-def is_torch_complex_dtype(tensor: TensorLike) -> bool:
+def is_torch_complex_dtype(tensor_or_dtype: TensorLike | torch.dtype) -> bool:
```
Comment (justinchuby) [inline torch/onnx/_internal/fx/type_utils.py:33]:
```suggestion
def is_torch_complex_dtype(tensor_or_dtype: torch.dtype) -> bool:
```

File: torch/onnx/_internal/fx/type_utils.py [hunk 13]
```diff
@@ -30,9 +30,11 @@ def dtype(self) -> Optional[torch.dtype]:
...
     # NOTE: This is needed as TorchScriptTensor is nor supported by torch.is_complex()
-    return tensor.dtype in _COMPLEX_TO_FLOAT
+    if isinstance(tensor_or_dtype, torch.dtype):
+        return tensor_or_dtype in _COMPLEX_TO_FLOAT
+    return tensor_or_dtype.dtype in _COMPLEX_TO_FLOAT
```
Comment (justinchuby) [inline torch/onnx/_internal/fx/type_utils.py:37]:
```suggestion
    return tensor_or_dtype in _COMPLEX_TO_FLOAT
```

to reduce confusion? i.e. Users don't need to consider whether to use `is_torch_complex_dtype(val)` or `is_torch_complex_dtype.dtype`. Ideally there should only be one correct way.

**Comment by justinchuby** [comment]
LGTM. Just a few minor comments

File: torch/onnx/_internal/fx/onnxfunction_dispatcher.py (hunk 11 above)
Comment (BowenBao) [reply torch/onnx/_internal/fx/onnxfunction_dispatcher.py:143]: Not really. Each op will populate two diagnostics of this same rule, but one for this function and another for `get_function_overloads`.

Comment (titaiwangms) [reply torch/onnx/_internal/fx/onnxfunction_dispatcher.py]: Moved it down to `get_function_overloads`

File: test/onnx/onnx_test_common.py [hunk 14]
```diff
@@ -395,11 +402,17 @@ def _compare_pytorch_onnx_with_ort(
...
+
     if len(ref_outputs) != len(ort_outputs):
         raise AssertionError(
             f"Expected {len(ref_outputs)} outputs, got {len(ort_outputs)}"
         )
     for ref_output, ort_output in zip(ref_outputs, ort_outputs):
+        # NOTE: ONNX Runtime doesn't support tensor of complex64/complex128, so we
+        # convert them to float32/float64.
+        # TODO: Need stft enabled to test complex64/complex128
+        if isinstance(ref_output, torch.Tensor) and torch.is_complex(ref_output):
```
Comment (BowenBao) [inline test/onnx/onnx_test_common.py:414]: You should add a new input/output adaptor step for this. Checkout `RemoveNonTensorInputStep` for example.
Feel free to do it in this or follow-up PR.

File: test/onnx/onnx_test_common.py [hunk 15]
```diff
@@ -344,9 +344,16 @@ def run_ort(
...
-    )
+    # NOTE: ONNX doesn't support tensor of complex64/complex128, so we
+    # convert them to float32/float64.
+    pytorch_inputs = [
+        torch.view_as_real(input)
+        if isinstance(input, torch.Tensor) and torch.is_complex(input)
+        else input
+        for input in pytorch_inputs
+    ]
+    ort_input = {k: v.cpu().numpy() for k, v in zip(input_names, pytorch_inputs)}
```
Comment (BowenBao) [inline test/onnx/onnx_test_common.py:355]: ditto: input adapter step

**Review by BowenBao — APPROVED** [review]
LGTM w/ comments. Please also resolve others' comments before landing.

File: test/onnx/test_fx_op_consistency.py (hunk 6 above)
Comment (titaiwangms) [reply test/onnx/test_fx_op_consistency.py:650]: Found it's strict. So I added tasks to https://github.com/pytorch/pytorch/issues/105338. But it's doable.

File: test/onnx/onnx_test_common.py (hunk 14 above)
Comment (titaiwangms) [reply test/onnx/onnx_test_common.py]: Done

File: test/onnx/onnx_test_common.py (hunk 15 above)
Comment (titaiwangms) [reply test/onnx/onnx_test_common.py:347]: Done

**Comment by justinchuby** [comment]
Are we handling any complex constants that may be inside the graph? Maybe that’s an unlikely case that we don’t need to care yet?

**Comment by titaiwangms** [comment]
> Are we handling any complex constants that may be inside the graph? Maybe that’s an unlikely case that we don’t need to care yet?

It's not covered in this PR. We can add it when we have a model or use case.

**Pull request closed without merging**